Read Sphinx output and create struct

anoshaminai opened the pull request:
@sandlerben @roornelas 

This PR is an extension of #26 - the function to run the command is now fixed in this PR, but I can also fix it in #26 if that's less confusing. There are two main new things: 1) StartTranscription runs a command and returns a string which is the output file name where the output is stored and 2) the func transcriptionOutputToStruct was created which reads a file (with the output) and creates a go struct.

Commit:
start transcription job

Commit:
Merge branch 'master' of https://github.com/hack4impact/audio-transcription-service into run-sphinx-command-line

Merging in remote changes

roornelas committed:
config working

roornelas committed:
updated with PR comments

roornelas committed:
dependencies

roornelas committed:
removed error printing

Commit:
pulled git changes and merged

sandlerben committed:
Merge branch 'new-config' of https://github.com/hack4impact/transcribe4all into ibm-websockets

sandlerben committed:
Refactored ibm transcription code

Commit:
merged in sphinx code

Commit:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into run-sphinx-command-line

Commit:
fixed start transcription method from before

roornelas committed:
small fix

roornelas committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into new-config

roornelas committed:
Merge pull request #30 from hack4impact/new-config

config working

Commit:
added transcription struct

Commit:
Merge pull request #35 from hack4impact/form-ui

Basic UI for starting a job

Commit:
Added transc to output func

Commit:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into anosha-read-json-file-to-struct

Commit:
updated godeps

Commit:
separated StartTransc and transcripToOutput

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into ibm-websockets

Commit:
Form uses Semantic UI and JQuery, which are included in /static

sandlerben committed:
Updated ibm transcription to use gorilla websockets; now working

Commit:
Jquery needed

Commit:
Favicon, webpage need icon

Commit:
form field names changed, so names must be updated for form parsing

Commit:
SemanticUI login form example, applied to our form, with favicon, title, footer styled in H4I blue

Commit:
Deleted jquery dir because only jquery min file needed, not migrate

Commit:
Added search word field; changed icons; appended data structure and changed MakeTaskFunction arguments

sandlerben committed:
ibm transcription is working

Commit:
URL, email validation, not just must have value

Commit:
Deleted unnecessary semanticUI files

sandlerben commented on transcription/transcription.go:
```diff
@@ -38,6 +45,49 @@ func msgHeaders(from string, to []string, subject string) string {
 
+// StartTranscription transcribes a given file using Sphinx.
+// File name should be in "name.wav" format.
+func StartTranscription(fileName string, command string) (string, error) {
```
Rename to `StartSphinxTranscription` or even better `SphinxTranscription`

yoninachmany replied:
bump

sandlerben commented on transcription/transcription.go:
```diff
@@ -38,6 +45,49 @@ func msgHeaders(from string, to []string, subject string) string {
+// File name should be in "name.wav" format.
+func StartTranscription(fileName string, command string) (string, error) {
+
+	cmd := exec.Command("java", "-jar", command, fileName)
```
Shouldn't you run `./gradlew`? Definitely try out your code to make sure it works (e.g. add a call to your function in `main.main`.

yoninachmany replied:
Where does this stand?

sandlerben commented on transcription/transcription.go:
```diff
@@ -38,6 +45,49 @@ func msgHeaders(from string, to []string, subject string) string {
+func StartTranscription(fileName string, command string) (string, error) {
+
+	cmd := exec.Command("java", "-jar", command, fileName)
+	_, err := cmd.CombinedOutput()
```
If you aren't using the output why not use [`cmd.Run`](https://golang.org/pkg/os/exec/#Cmd.Run)

yoninachmany replied:
bump

sandlerben commented on transcription/transcription.go:
```diff
@@ -38,6 +45,49 @@ func msgHeaders(from string, to []string, subject string) string {
+	if err != nil {
+		return "", err
+	}
+	outputFileName := "/Sphinx/files/" + fileName + "-json.txt"
```
Why not just `.json`?

anoshaminai replied:
It is output to a text file.

yoninachmany replied:
@sandlerben how does that sound?

sandlerben replied:
`.json` is a ascii text format. I would use that but it's not a big deal either way.

sandlerben commented on transcription/transcription.go:
```diff
@@ -38,6 +45,49 @@ func msgHeaders(from string, to []string, subject string) string {
+	// if _, err := transcriptionOutputToStruct(outputFileName); err != nil {
+	// 	return err
+	// }
+	return outputFileName, nil
```
Shouldn't you return the transcription output and not the file name?

anoshaminai replied:
The transcription output is a file which is stored at that filename location - maybe it would make sense to rename the variable to 'outputFileLoc' or 'outputFile'? The second method (transcriptionOutputToStruct) reads the output file and creates a struct. One option would be for this method to call transcriptionOutputToStruct and return a struct. I think that makes a lot of sense.

yoninachmany replied:
@sandlerben thoughts?

sandlerben replied:
`outputPath`?

sandlerben commented on transcription/transcription.go:
```diff
@@ -38,6 +45,49 @@ func msgHeaders(from string, to []string, subject string) string {
+	nText := bytes.IndexByte(bytesText, 0)
+	nMeta := bytes.IndexByte(bytesMeta, 0)
+	sText := string(bytesText[:nText])
+	sMeta := string(bytesText[:nMeta])
```
What is this for? Why don't you use the built-in json package?

anoshaminai replied:
Which built in json package? I am reading in the data from a text file, which is how Sphinx stores it. I'm reading each line and storing the data in the appropriate struct fields .

yoninachmany replied:
For reference, built-in json package: https://golang.org/pkg/encoding/json/

Not sure what the situation here is, but would generally recommend using existing json parsing library.

sandlerben committed:
Fix travis to ignore vendors

sandlerben committed:
Read result into struct; Stop websocket from timing out

sandlerben committed:
Now using ws.ReadJSON function

sandlerben committed:
Return a pointer of a IBMResult for efficiency reasons

sandlerben committed:
Handle all errors

Commit:
Semantic tokenfield files for gmail-like word separation

Commit:
Links for semantic tokenfield src and ids for divs

Commit:
Merge pull request #38 from hack4impact/style-form

Styled form using SemanticUI; added search word field

sandlerben committed:
Added GetTranscript function

sandlerben committed:
Added ibm tests

sandlerben committed:
Updated godeps

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into ibm-websockets

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into ibm-websockets

roornelas committed:
Merge pull request #31 from hack4impact/rodrigo-sphinx

Rodrigo sphinx

Commit:
changed fun name

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into hook-app-together

sandlerben committed:
Moved config to its own package. Refactored and renamed other code

yoninachmany commented:
@anoshaminai bump on making sure conflicts with master are resolved

sandlerben committed:
Added logging and fixed a few bugs to get full transcription working

sandlerben committed:
A few small things

sandlerben committed:
Merge pull request #39 from hack4impact/ibm-websockets

Use client-side websockets to call IBM Speech to Text

sandlerben committed:
Updated vendor

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all into hook-app-together

sandlerben committed:
Added a hacky version of flashes to the app

sandlerben committed:
Removed dead code and fully migrated logging to logrus

sandlerben committed:
Merge pull request #43 from hack4impact/hook-app-together

Hook app together and add logging

sandlerben committed:
Annotate errors with stack trace information

sandlerben committed:
Merge pull request #47 from hack4impact/error-context

Annotate errors with stack trace information

sandlerben committed:
Added actual sessions

sandlerben committed:
Merge pull request #48 from hack4impact/real-app-flashing

Added actual sessions

sandlerben committed:
Errors are handled with email now

sandlerben committed:
Merge pull request #49 from hack4impact/email-upon-failure

Errors are handled with email now

sandlerben committed:
Updated completion message

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all

sandlerben committed:
Task executer now expires task information to avoid memory leaks

sandlerben committed:
File path is unique

sandlerben committed:
Merge pull request #50 from hack4impact/tasks-memory-leak

Task executer now expires task information to avoid memory leaks

sandlerben committed:
Updated vendors

sandlerben committed:
Merge branch 'master' of https://github.com/hack4impact/transcribe4all

Commit:
changed to single method call required

Commit:
Fixed command to run transcription

Commit:
Used JSON package to read file to struct

Commit:
Merge conflicts

anoshaminai commented:
@sandlerben Addressed comments, also I thought I addressed conflicts but apparently not. When I call 'git merge origin/master' it says everything is up to date - what else do I need to do?

hack4impactadmin commented:
Did you fetch first?

On Thu, May 19, 2016, 7:35 PM Anosha Minai <notifications@github.com> wrote:

> @sandlerben <https://github.com/sandlerben> Addressed comments, also I
> thought I addressed conflicts but apparently not. When I call 'git merge
> origin/master' it says everything is up to date - what else do I need to do?
>
> —
>
>
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/hack4impact/transcribe4all/pull/36#issuecomment-220477418>
>
> --
> You received this message because you are subscribed to the Google Groups
> "hack4impact-exec-2016-17" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to hack4impact-exec-2016-17+unsubscribe@googlegroups.com.
> To post to this group, send email to
> hack4impact-exec-2016-17@googlegroups.com.
> To view this discussion on the web visit
> https://groups.google.com/d/msgid/hack4impact-exec-2016-17/hack4impact/transcribe4all/pull/36/c220477418%40github.com
> <https://groups.google.com/d/msgid/hack4impact-exec-2016-17/hack4impact/transcribe4all/pull/36/c220477418%40github.com?utm_medium=email&utm_source=footer>
> .
> For more options, visit https://groups.google.com/d/optout.
>

sandlerben commented:
Also I think the account situation is a little messed up I'll fix it later.

On Thu, May 19, 2016, 8:35 PM hack4impactadmin <notifications@github.com>
wrote:

> Did you fetch first?
>
> On Thu, May 19, 2016, 7:35 PM Anosha Minai <notifications@github.com>
> wrote:
>
> > @sandlerben <https://github.com/sandlerben> Addressed comments, also I
> > thought I addressed conflicts but apparently not. When I call 'git merge
> > origin/master' it says everything is up to date - what else do I need to
> do?
> >
> > —
> >
> >
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/hack4impact/transcribe4all/pull/36#issuecomment-220477418
> >
> >
> > --
> > You received this message because you are subscribed to the Google Groups
> > "hack4impact-exec-2016-17" group.
> > To unsubscribe from this group and stop receiving emails from it, send an
> > email to hack4impact-exec-2016-17+unsubscribe@googlegroups.com.
> > To post to this group, send email to
> > hack4impact-exec-2016-17@googlegroups.com.
> > To view this discussion on the web visit
> >
> https://groups.google.com/d/msgid/hack4impact-exec-2016-17/hack4impact/transcribe4all/pull/36/c220477418%40github.com
> > <
> https://groups.google.com/d/msgid/hack4impact-exec-2016-17/hack4impact/transcribe4all/pull/36/c220477418%40github.com?utm_medium=email&utm_source=footer
> >
> > .
> > For more options, visit https://groups.google.com/d/optout.
> >
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/hack4impact/transcribe4all/pull/36#issuecomment-220488888>
>

anoshaminai commented:
@sandlerben Okay all fixed! I moved the Sphinx stuff into a sphinx.go folder, I need to add the MakeSphinxTaskFunction method but I will do that in a separate PR. Let me know what you think and then I'll merge.

Commit:
Changed transcription.go to sphinx.go

sandlerben commented on transcription/sphinx.go:
```diff
@@ -0,0 +1,57 @@
+
+// SphinxTranscription transcribes a given file using Sphinx.
+// File name should not include the type extension.
+func SphinxTranscription(fileName string) (Transcription, error) {
```
Change to `TranscribeWithShinx`?

sandlerben replied:
~~`fileName`~~ -> `filePath` to be consistent with other code

sandlerben commented on transcription/sphinx.go:
```diff
@@ -0,0 +1,57 @@
+
+//Transcription contains the transcription text and metadata of a transcription
+//job
+type Transcription struct {
```
We probably want to use the same `TranscriptionResult` struct for IBM and Sphinx right? If so, I think we should further refactor so that `TranscriptionResult` is defined in `transcription.go`. We can move the `MakeXTaskFunction`s from `utils.go` to `transcription.go`, too.

sandlerben commented on transcription/sphinx.go:
```diff
@@ -0,0 +1,57 @@
+// SphinxTranscription transcribes a given file using Sphinx.
+// File name should not include the type extension.
+func SphinxTranscription(fileName string) (Transcription, error) {
+	var result Transcription
```
Nit: this function could probably use one or two spaces to make it easier on the eyes

sandlerben commented on transcription/sphinx.go:
```diff
@@ -0,0 +1,57 @@
+	return jsonData, nil
+}
+
+//Need to add a corresponding MakeSphinxTaskFunction
```
You can remove this comment once you add the task function

sandlerben replied:
But let's do that before merging

sandlerben commented on transcription/transcription_test.go:
```diff
@@ -1,130 +0,0 @@
-package transcription
```
Why did you delete this?

sandlerben commented:
Added a few comments, besides that lgtm
